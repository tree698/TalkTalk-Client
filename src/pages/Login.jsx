import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import toast from 'react-hot-toast';
import { useApiContext } from '../context/ApiContext';

export default function Login() {
  const [loginInfo, setLoginInfo] = useState({ username: '', password: '' });
  const [error, setError] = useState();

  const navigate = useNavigate();
  const { logIn } = useApiContext();

  // useEffect(() => {
  //   error && toast.error(error);
  // }, [error]);

  const handleSubmit = (event) => {
    event.preventDefault();
    logIn(loginInfo.username, loginInfo.password); //
    // .catch((error) => setError((prev) => error.toString()));
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setLoginInfo({ ...loginInfo, [name]: value });
  };

  const INPUT_STYLE =
    'w-full px-1 md:px-2 py-1 md:py-2 text-sm md:text-lg lg:text-base outline-none border border-superLightGray rounded mt-1 md:mt-2 mb-2 md:mb-4 placeholder:italic placeholder:text-superLightGray placeholder:text-xs md:placeholder:text-sm';

  return (
    <section className="basis-1/2 lg:basis-2/5 w-full flex justify-center items-center bg-signupLoginBG ">
      <div className="w-full px-14 lg:px-20">
        <div className="flex justify-end items-center">
          <p className="text-[10px] md:text-xs">Don't you have an account?</p>
          <button
            className="py-1 px-1 md:px-2 ml-2 md:ml-3 text-[10px] md:text-xs border border-superLightGray rounded-md hover:bg-accent hover:text-white transition-all delay-150 duration-300 ease-in-out"
            onClick={() => navigate('/signup')}
            data-testid="signupBtn"
          >
            SIGN UP
          </button>
        </div>

        <h1 className="font-bold text-xl md:text-2xl lg:text-3xl mt-4 md:mt-6 lg:mt-8 mb-1 ">
          Welcome Back
        </h1>
        <p className="text-[10px] md:text-x mb-4 md:mb-6 lg:mb-8">
          Login your account
        </p>

        <form onSubmit={handleSubmit} role="form">
          <div>
            <label
              htmlFor="username"
              className="text-xs md:text-sm lg:text-base"
            >
              Username
            </label>
            <br />
            <input
              type="text"
              name="username"
              value={loginInfo.username ?? ''}
              onChange={handleChange}
              placeholder="Your username"
              id="username"
              required
              className={`${INPUT_STYLE}`}
            />
            <br />
            <label
              htmlFor="password"
              className="text-xs md:text-sm lg:text-base"
            >
              Password
            </label>
            <br />
            <input
              type="password"
              name="password"
              value={loginInfo.password ?? ''}
              onChange={handleChange}
              placeholder="Your password"
              id="password"
              required
              className={`${INPUT_STYLE}`}
            />
            <br />
          </div>
          <button
            data-testid="loginBtn"
            className="w-full lg:w-2/5 text-sm md:text-lg lg:text-base py-1 md:py-2 mt-4  text-white bg-accent rounded-3xl transition-all delay-150 duration-300 ease-in-out hover:font-bold hover:scale-105 hover:shadow-xl"
          >
            Login
          </button>
        </form>
      </div>
    </section>
  );
}
