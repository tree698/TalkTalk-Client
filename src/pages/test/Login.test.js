import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { withApiContext, withRouter } from '../../test/utils';
import Login from '../Login';

describe('Upload', () => {
  const fakeContext = {
    logIn: jest.fn(),
  };

  afterEach(() => fakeContext.logIn.mockReset());

  it('renders correctly', () => {
    const component = renderer.create(
      withApiContext(
        withRouter(<Route path="/" element={<Login />} />),
        fakeContext
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('display input value correctly', async () => {
    render(
      withApiContext(
        withRouter(<Route path="/" element={<Login />} />),
        fakeContext
      )
    );
    const username = screen.getByRole('textbox', { name: /username/i });

    userEvent.type(username, 'ellie');

    await waitFor(() => {
      expect(username).toHaveValue('ellie');
    });
  });

  it('navigate to signup page when clicking the signup button', async () => {
    render(
      withApiContext(
        withRouter(
          <>
            <Route path="/login" element={<Login />} />
            <Route path="/signup" element={<p>Sign</p>} />
          </>,
          '/login'
        ),
        fakeContext
      )
    );

    const button = screen.getByTestId('signupBtn');
    userEvent.click(button);

    await waitFor(() => {
      expect(screen.getByText('Sign')).toBeInTheDocument();
    });
  });

  it('Successfully login', async () => {
    fakeContext.logIn.mockImplementation((username, password) => {
      return 'Successfully login';
    });

    render(
      withApiContext(
        withRouter(<Route path="/" element={<Login />} />),
        fakeContext
      )
    );

    const form = screen.getByTestId('loginBtn');
    userEvent.click(form);

    await waitFor(() => {
      expect(fakeContext.logIn).toBeCalled();
    });
  });
});
