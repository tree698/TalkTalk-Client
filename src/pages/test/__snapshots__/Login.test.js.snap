// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload renders correctly 1`] = `
<section
  className="basis-1/2 lg:basis-2/5 w-full flex justify-center items-center bg-signupLoginBG "
>
  <div
    className="w-full px-14 lg:px-20"
  >
    <div
      className="flex justify-end items-center"
    >
      <p
        className="text-[10px] md:text-xs"
      >
        Don't you have an account?
      </p>
      <button
        className="py-1 px-1 md:px-2 ml-2 md:ml-3 text-[10px] md:text-xs border border-superLightGray rounded-md hover:bg-accent hover:text-white transition-all delay-150 duration-300 ease-in-out"
        data-testid="signupBtn"
        onClick={[Function]}
      >
        SIGN UP
      </button>
    </div>
    <h1
      className="font-bold text-xl md:text-2xl lg:text-3xl mt-4 md:mt-6 lg:mt-8 mb-1 "
    >
      Welcome Back
    </h1>
    <p
      className="text-[10px] md:text-x mb-4 md:mb-6 lg:mb-8"
    >
      Login your account
    </p>
    <form
      onSubmit={[Function]}
      role="form"
    >
      <div>
        <label
          className="text-xs md:text-sm lg:text-base"
          htmlFor="username"
        >
          Username
        </label>
        <br />
        <input
          className="w-full px-1 md:px-2 py-1 md:py-2 text-sm md:text-lg lg:text-base outline-none border border-superLightGray rounded mt-1 md:mt-2 mb-2 md:mb-4 placeholder:italic placeholder:text-superLightGray placeholder:text-xs md:placeholder:text-sm"
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="Your username"
          required={true}
          type="text"
          value=""
        />
        <br />
        <label
          className="text-xs md:text-sm lg:text-base"
          htmlFor="password"
        >
          Password
        </label>
        <br />
        <input
          className="w-full px-1 md:px-2 py-1 md:py-2 text-sm md:text-lg lg:text-base outline-none border border-superLightGray rounded mt-1 md:mt-2 mb-2 md:mb-4 placeholder:italic placeholder:text-superLightGray placeholder:text-xs md:placeholder:text-sm"
          id="password"
          name="password"
          onChange={[Function]}
          placeholder="Your password"
          required={true}
          type="password"
          value=""
        />
        <br />
      </div>
      <button
        className="w-full lg:w-2/5 text-sm md:text-lg lg:text-base py-1 md:py-2 mt-4  text-white bg-accent rounded-3xl transition-all delay-150 duration-300 ease-in-out hover:font-bold hover:scale-105 hover:shadow-xl"
        data-testid="loginBtn"
      >
        Login
      </button>
    </form>
  </div>
</section>
`;
